### PR TITLE
mnistを学習できるようにした

### PR DESCRIPTION
**Pull request by Kyosuke1004**
どこまで作り込めばよいかわからず，一番わかりやすいかなと思ったサイトを参考にしました

### CODE REVIEW

File: scripts/networks.py [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+        self.fc1 = torch.nn.Linear(28 * 28, 1000)
+        self.fc2 = torch.nn.Linear(1000, 10)
+        
+    # 順伝搬
+    def forward(self, x):
+        x = self.fc1(x)
+        x = torch.sigmoid(x)
+        x = self.fc2(x)
+        
+        return f.log_softmax(x, dim=1)
```
Comment (uchikun2493) [inline scripts/networks.py]: log_softmax()は不要です
誤差計算（cross_entropy_loss）で内包できます

File: scripts/networks.py [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# import modules
+import torch
+import torch.nn.functional as f
+
+class MyNet(torch.nn.Module):
```
Comment (uchikun2493) [inline scripts/networks.py]: ネットワーク名は一般に使える命名が良いと思います，例えばClassifier，MNISTClassifier，MLPClassifier，MLPCls,
とか

File: scripts/networks.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
+# import modules
+import torch
+import torch.nn.functional as f
```
Comment (uchikun2493) [inline scripts/networks.py]: 細かいけど，Fで略すことが多いです
```python
import torch.nn.functional as F
```

File: scripts/mnist_mlp.py [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+            # 全結合のみのネットワークでは入力を1次元に
+            # print(data.shape)
+            # torch.Size([128, 1, 28, 28])
+            data = data.view(-1, 28 * 28)
+            # print(data.shape)
+            # torch.Size([128, 784])
+            
+            optimizer.zero_grad()
+            output = net(data)
+            loss = f.nll_loss(output, target)
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: 誤差計算は事前にcriterionを定義して行わせましょう
学習ループに入る前に以下を定義します
```python
criterion = nn.CrossEntropyLoss()
```

Comment (uchikun2493) [inline scripts/mnist_mlp.py]: この辺の並び方は他のリポジトリを参考にまとまった並びにしてほしい

File: scripts/mnist_mlp.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+            loss.backward()
+            optimizer.step()
+            
+            if i % 10 == 0:
+                print('Training log: {} epoch ({} / 60000 train. data). Loss: {}'.format(epoch + 1,
+                                                                                         (i + 1) * 128,
+                                                                                         loss.item())
+                      )
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: 10回のイテレーションごとだと多いと思うので1エポッごとに誤差を出してほしいです〜

**Review by uchikun2493 — COMMENTED** [review]
まとめました，他気づいたらまた書きます

File: scripts/mnist_mlp.py [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+
+from torch.utils.data import DataLoader
+from torchvision import datasets, transforms
+
+import networks
+
+# NNISTをロードする関数
+def load_MNIST(batch=128, intensity=1.0):
+    train_loader = torch.utils.data.DataLoader(
+        datasets.MNIST('./data',
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: `./data`の指定だと`scripts/data/...`の中にデータが入りますが，今回は`../data`ディレクトリがあるので，そこを指定しましょう
そしてscriptsの中にダウンロードされたデータは不要なので削除をお願いします

File: scripts/mnist_mlp.py [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+    
+    # 学習結果の保存用
+    history = {
+        'train_loss': [],
+        'test_loss': [],
+        'test_acc': [],
+    }
+    
+    # ネットワークを構築
+    net: torch.nn.Module = networks.MyNet()
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: 書き方としては
```python
net = networks.MyNet()
```
でOKです

File: scripts/mnist_mlp.py [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+    
+    # -------------------------------------
+    # モデルの学習
+    # -------------------------------------
+    
+    # 学習，テスト用ループ
+    for epoch in range(epochs):
+        
+        # 学習パート
+        loss = None
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: 学習パートというコメントの下に`loss = None`は不自然な気がします

File: scripts/mnist_mlp.py [hunk 9]
```diff
@@ -0,0 +1,135 @@
+# import modules
+import torch
+import torch.nn.functional as f
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: functionalはcriterionを定義することで最終的に不要となります

File: scripts/mnist_mlp.py [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+
+    test_loader = torch.utils.data.DataLoader(
+        datasets.MNIST('./data',
+                       train=False,
+                       transform=transforms.Compose([
+                           transforms.ToTensor(),
+                           transforms.Lambda(lambda x: x * intensity)
+                       ])),
+        batch_size=batch,
+        shuffle=True)
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: testの方では`shuffle`をTrueにする必要はありません

File: scripts/mnist_mlp.py [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+        
+        # 学習パート
+        loss = None
+        # 学習開始
+        net.train(True)
+        for i, (data, target) in enumerate(loaders['train']):
+            # 学習部分
+            # 全結合のみのネットワークでは入力を1次元に
+            # print(data.shape)
+            # torch.Size([128, 1, 28, 28])
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: このあたりの不要なコメントは削除しておいてください

**Review by uchikun2493 — COMMENTED** [review]
コメント追加しました

**Comment by uchikun2493** [comment]
- to.device(device)でGPUを使った学習ができるようにしてほしいです
- モデルの変数名を`net`にしてますが`model`にしてほしいです（他のリポジトリとの整合性のため）

File: scripts/mnist_mlp.py [hunk 12]
```diff
@@ -0,0 +1,138 @@
...
+        for inputs, labels in loaders['train']:
+            # 指定したdeviceに変数を変換
+            inputs = inputs.to(device)
+            labels = labels.to(device)
+            # データの整形
+            inputs = inputs.view(-1, 28 * 28)
+            # ニューラルネットワークへのデータ入力
+            outputs = model(inputs)
+            # 誤差を算出
+            loss = criterion(outputs, labels)
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: train用のlossを用意する方がいいかもしれません
```python
train_loss = 0.0
for ...:
    # 処理
    train_loss = loss.item
train_loss /= data_num

# ...省略

history['train_loss'].append(train_loss)
```

File: scripts/mnist_mlp.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+        # 平均誤差を算出
+        test_loss /= data_num
+        correct /= data_num
+        # 1エポックあたりの誤差と正解率を表示
+        print('epoch{} Test loss (avg): {}, Accuracy: {}'.format(epoch + 1, test_loss, correct))
+
+        # 誤差と正解率を代入
+        history['train_loss'].append(loss.item())
+        history['test_loss'].append(test_loss)
+        history['test_acc'].append(correct)
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: testと比較もしたいので，trainのaccuracyも計算してほしいです！

File: scripts/mnist_mlp.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+    plt.plot(range(1, epochs + 1), history['test_loss'], label='test_loss')
+    plt.xlabel('epoch')
+    plt.legend()
+    plt.savefig('../outputs/loss.png')
+ 
+    plt.figure()
+    plt.plot(range(1, epochs + 1), history['test_acc'])
+    plt.title('test accuracy')
+    plt.xlabel('epoch')
+    plt.savefig('../outputs/test_acc.png')
```
Comment (uchikun2493) [inline scripts/mnist_mlp.py]: trainのaccuracyを追加したら，testと合わせて描画してほしいです
保存名はscc.pngにしますか
